liota/core/package_manager.py

```diff
@@ -30,4 +30,1104 @@
+# May raise IOError upon "open"
+
+def sha1sum(path_file):
+    sha1 = hashlib.sha1()
+    with open(path_file, "rb") as fp:
+        while True:
+            data = fp.read(65536) # buffer size
+            if not data:
+                break
+            sha1.update(data)
+    return sha1
+
+class PackageRecord:
+    """
+    PackageRecord is instantiated for each package loaded.
+    It contains necessary information of packages, including but may not be
+    limited to file names, checksums, and refs to package class instances.
+    These records make it possible to start and stop package class instances
+    elegantly.
+    """
+
+    def __init__(self, file_name):
+        self._file_name = file_name
+        self._ext = None
+        self._sha1 = None
+        self._dependents = {} # key: dependent name, value: None
+        self._dependencies = []
+
+        #-------------------------------------------------------------------
+        # To guarantee successful garbage collection when record is removed,
+        # this should be the ONLY variable to keep its reference.
+
+        self._instance = None
+
+    def set_instance(self, obj):
+        if self._instance is not None:
+            log.warning("Should not override instance of package class")
+            return False
+        self._instance = obj
+        return True
+
+    def get_instance(self):
+        return self._instance
+
+    def set_sha1(self, sha1):
+        self._sha1 = sha1
+
+    def get_sha1(self):
+        return self._sha1
+
+    def set_ext(self, ext):
+        self._ext = ext
+
+    def get_ext(self):
+        return self._ext
+
+    def get_dependents(self):
+        return self._dependents.keys()
+
+    def add_dependent(self, file_name):
+        self._dependents[file_name] = None
+
+    def del_dependent(self, file_name):
+        del self._dependents[file_name]
+
+    def get_dependencies(self):
+        return self._dependencies
+
+    def set_dependencies(self, list_dependencies):
+        self._dependencies = list_dependencies
+
+class PackageThread(Thread):
+    """
+    PackageThread should be instantiated only once.
+    Its instance serves core functionalities of Liota package manager,
+    including load/unload of packages, maintenance of loaded package records,
+    etc.
+    """
+
+    def __init__(self, name=None):
+        Thread.__init__(self, name=name)
+
+        global package_path
+
+        self._packages_loaded = {} # key: package name, value: PackageRecord obj
+        self._resource_registry = ResourceRegistry()
+        self._resource_registry.register("package_conf", package_path)
+        self.flag_alive = True
+        self.start()
+
+    #-----------------------------------------------------------------------
+    # This method is used to handle listing commands
+
+    def _cmd_handler_list(self, parameter):
+        if parameter == "packages" or parameter == "pkg":
+            log.warning("List of packages - \n\t%s" \
```

Comment:
we do want to these printed out in liota.log since this is result of these cmds (e.g., list pkg...). 
Our default message level is warning, so  better to be warning.